app/src/main/java/org/mozilla/fenix/session/NotificationSessionObserver.kt

```diff
@@ -5,38 +5,46 @@
 package org.mozilla.fenix.session
 
 import android.content.Context
-import mozilla.components.browser.session.Session
-import mozilla.components.browser.session.SessionManager
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.map
+import mozilla.components.browser.state.selector.privateTabs
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
 import org.mozilla.fenix.ext.components
-import org.mozilla.fenix.ext.sessionsOfType
 
 /**
  * This observer starts and stops the service to show a notification
  * indicating that a private tab is open.
  */
-
 class NotificationSessionObserver(
     private val context: Context,
     private val notificationService: SessionNotificationService.Companion = SessionNotificationService
-) : SessionManager.Observer {
-
-    override fun onSessionRemoved(session: Session) {
-        val privateTabsEmpty = context.components.core.sessionManager.sessionsOfType(private = true).none()
-
-        if (privateTabsEmpty) {
-            notificationService.stop(context)
+) {
+
+    private var scope: CoroutineScope? = null
+    private var started = false
+
+    @ExperimentalCoroutinesApi
+    fun start() {
+        scope = context.components.core.store.flowScoped { flow ->
+            flow.map { state -> state.privateTabs.isNotEmpty() }
+                .ifChanged()
+                .collect { hasPrivateTabs ->
+                    if (hasPrivateTabs) {
+                        notificationService.start(context)
+                        started = true
+                    } else if (started) {
+                        notificationService.stop(context)
+                        started = false
+                    }
+                }
         }
     }
 
-    override fun onAllSessionsRemoved() {
-        notificationService.stop(context)
-    }
-
-    override fun onSessionAdded(session: Session) {
-        // Custom tabs are meant to feel like part of the app that opened them, not Fenix, so we
-        // don't need to show a 'close tab' notification for them
-        if (session.private && !session.isCustomTabSession()) {
-            notificationService.start(context)
-        }
+    fun stop() {
+        scope?.cancel()
```

Comment:
Quick Q: Is it a possible problem that this scope cancel is never called AFAICT? It's not lifecycle aware and we are manually starting it in `FenixApplication` cc @NotWoods @csadilek

Reply:
Should be fine since we want the notification to stick around

Reply:
We have a few store observers / scopes tied to the whole application lifecycle e.g. `MigrationTelemetryListener`, `MigrationPushRenewer`. Those never get stopped and don't need to. I guess we could've not assigned the scope and removed the `stop` method, but yes, it should not be a problem.

Reply:
@ekager @NotWoods Actually, we do have a problem here after all! The `NotificationSessionObserver` is leaking the `HomeActivity` as it's initialized in `HomeActivity.onCreate` and uses it as a context. When the `HomeActivitiy` is re-created, we never cancel the old scope and therefore leak the old `HomeActivity`. You can reproduce this by enabling "Don't keep activities"...then leak canary gets pretty mad at us :).

@NotWoods I saw you were working on an uplift of this class so we can fix it there? Let's chat in our PR...